physics/SFC_Models/Land/Noahmp/noahmpdrv.F90

```diff
@@ -9,99 +9,469 @@
+        print *, 'FATAL ERROR in noahmpdrv_timestep_init: problem in set_soilveg_noahmp'
+        errmsg = 'FATAL ERROR in noahmpdrv_timestep_init: problem in set_soilveg_noahmp'
+        return
+  endif
+
+  n_stc = 0
+  n_slc = 0
+  !!do moisture/temperature adjustment for consistency after increment add
+  if (Land_IAU_Control%do_stcsmc_adjustment) then
+    if (Land_IAU_Control%upd_stc) then
+      do i=1,lensfc
+        if (stc_updated(i) == 1 ) then ! soil-only location
+            n_stc = n_stc+1
+            soiltype = soiltyp(i)
+            do l = 1, lsoil_incr
+                !case 1: frz ==> frz, recalculate slc, smc remains
+                !case 2: unfrz ==> frz, recalculate slc, smc remains
+                !both cases are considered in the following if case
+                if (stc(i,l) .LT. tfreez )then
+                  !recompute supercool liquid water,smc_anl remain unchanged
+                  smp = hfus*(tfreez-stc(i,l))/(grav*stc(i,l)) !(m)
+                  slc_new=maxsmc(soiltype)*(smp/satpsi(soiltype))**(-1./bb(soiltype))
+                  slc(i,l) = max( min( slc_new, smc(i,l)), 0.0 )
+                endif
+                !case 3: frz ==> unfrz, melt all soil ice (if any)
+                if (stc(i,l) .GT. tfreez )then !do not rely on stc_bck
+                  slc(i,l)=smc(i,l)
+                endif
+            enddo
+        endif
+      enddo    
+    endif 
+
+    if (Land_IAU_Control%upd_slc) then
+      dz(1) = -zsoil(1)
+      do l = 2, km 
+          dz(l) = -zsoil(l) + zsoil(l-1) 
+      enddo 
+      ! print *, 'Applying soil moisture mins ' 
+      do i=1,lensfc
+        if (slc_updated(i) == 1 ) then 
+          n_slc = n_slc+1
+          ! apply SM bounds (later: add upper SMC limit)
+          do l = 1, lsoil_incr
+            ! noah-mp minimum is 1 mm per layer (in SMC)
+            ! no need to maintain frozen amount, would be v. small.
+            slc(i,l) = max( 0.001/dz(l), slc(i,l) )
+            smc(i,l) = max( 0.001/dz(l), smc(i,l) )
+          enddo
+        endif
+      enddo
+    endif
+  endif
 
-      end subroutine noahmpdrv_init
+    deallocate(stc_updated, slc_updated)
+    deallocate(mask_tile)
+  
+
+    write(*,'(a,i2)') ' statistics of grids with stc/smc updates for rank : ', Land_IAU_Control%me
+    write(*,'(a,i8)') ' soil grid total', lensfc
+    write(*,'(a,i8)') ' soil grid cells stc updated = ',nstcupd
+    write(*,'(a,i8)') ' soil grid cells slc updated = ',nslcupd
+    write(*,'(a,i8)') ' soil grid cells not updated, frozen = ',nfrozen
+    write(*,'(a,i8)') ' soil grid cells update, became frozen = ',nfrozen_upd
+    write(*,'(a,i8)') ' (not updated yet) snow grid cells = ', nsnowupd
+    write(*,'(a,i8)') ' grid cells, without soil or snow = ', nother 
+    write(*,'(a,i8)') ' soil grid cells with stc adjustment', n_stc 
+    write(*,'(a,i8)') ' soil grid cells with slc adjustment', n_slc 
+
+
+end subroutine noahmpdrv_timestep_init
+
+  !> \ingroup NoahMP_LSM
```

Comment:
This CCPP phase (timestep_finalize) subroutine can be removed since it is effectively empty.